csharp/src/Google.Protobuf/WellKnownTypes/Struct.cs

```diff
@@ -552,8 +604,13 @@ public enum KindOneofCase {
           }
         }
       }
+      pb::ProtoPreconditions.CheckMergedRequiredFields(this);
```

Comment:
It could be done that way in C++ since the merge method that is generated is a partial merge and is properly named for that. However in C# the merge method is not named for that. I could possibly make the generated MergeFrom and WriteTo methods not check for required fields and add an extension method that does check once (something like "CompleteMergeFrom" and "CompleteWriteTo"), but I thought it might be confusing to developers so I left it out.

Reply:
So do you want me to make the extensions or keep the duplicate calls?

Reply:
Is it possible to do like this? : 
1, Change the generated MergeFrom() to MergePartialFrom()
2, Add an "abstract class MessageBase : IMessage" or something for MergeFrom() that calls generated MergePartialFrom() and do the check. 
Then it will be same with C++

Note: I am not an expert in C#, my review is more based on protobuf features and other language implementations.

Reply:
Not exactly, since that would require changing the name of the method in the IMessage interface and would be an even bigger breaking change than just the addition of the IsInitialized method.

Reply:
I double checked the behaviors in other languages:
1, Serialized do not need to check required fields. (C++ is doing a debug check, not required)
2. Parse should check required fields, Merge should not check unless there a MergePaitialFrom method. 

So we can go like this: Not change the generated code's MergeFrom and WriteTo for check. Change the MessageParser to do the check.

Does this make sense to you?

Reply:
Yes, that's much easier